File: Java/src/main/java/testingwindows/ElementCountTrigger.java
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package testingwindows;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.coders.StringUtf8Coder;
import org.apache.beam.sdk.options.PipelineOptions;
import org.apache.beam.sdk.options.PipelineOptionsFactory;
import org.apache.beam.sdk.testing.TestStream;
import org.apache.beam.sdk.transforms.DoFn;
import org.apache.beam.sdk.transforms.GroupByKey;
import org.apache.beam.sdk.transforms.ParDo;
import org.apache.beam.sdk.transforms.windowing.*;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.TimestampedValue;
import org.joda.time.Duration;
import org.joda.time.Instant;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ElementCountTrigger {
    private static final Logger LOG = LoggerFactory.getLogger(ElementCountTrigger.class);

    public static void main(String[] args) {

        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(PipelineOptions.class);

        Pipeline p = Pipeline.create(options);

        Integer windowLength = 10;
        Integer allowLateSize = 5;
        Integer countElements = 2; // Trigger if 2 elements enter the window

        // Behaviour is explained in detail at the end of the file
        // Note: all elements would fall into the same window
        TestStream<String> streamEvents = TestStream.create(StringUtf8Coder.of())
                .addElements(
                        TimestampedValue.of("first", new Instant(0))
                )
                .advanceWatermarkTo(new Instant(2000))// move Watermark time to sec 2
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 2
                .addElements(
                        TimestampedValue.of("second", new Instant(1000))
                )
                .advanceWatermarkTo(new Instant(4000))// move Watermark time to sec 4
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 4
                .addElements(
                        TimestampedValue.of("third", new Instant(3000))
                )
                .advanceWatermarkTo(new Instant(6000)) // move Watermark time to sec 6
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 6
                .addElements(
                        // These elements arrive at the same processing time
                        TimestampedValue.of("same-time-1", new Instant(4000)),
                        TimestampedValue.of("same-time-2", new Instant(5000)),
                        TimestampedValue.of("same-time-3", new Instant(6000))
                )
                .advanceWatermarkTo(new Instant(9000))
                .advanceProcessingTime(Duration.standardSeconds(3))
                .addElements(
                        // These elements arrive at the same time
                        TimestampedValue.of("last-in-window", new Instant(8000))
                )
                .advanceWatermarkTo(new Instant(11000))
                .advanceProcessingTime(Duration.standardSeconds(2))
                .addElements(
                        // element arrives at sec 11 but fall in the window
                        TimestampedValue.of("Late-1", new Instant(0)),
                        TimestampedValue.of("Late-2", new Instant(5000)),
                        TimestampedValue.of("Late-3", new Instant(5550))
                )
                .advanceWatermarkTo(new Instant(12000))
                .advanceProcessingTime(Duration.standardSeconds(1))
                .addElements(
                        TimestampedValue.of("Late-4", new Instant(7000))

                )
                .advanceWatermarkTo(new Instant(13000))
                .advanceProcessingTime(Duration.standardSeconds(1))
                .addElements(
                        TimestampedValue.of("Late-5", new Instant(8000))

                )
                .advanceWatermarkTo(new Instant(14000))
                .advanceProcessingTime(Duration.standardSeconds(1))
                .addElements(
                        TimestampedValue.of("Late-6", new Instant(9000))

                )
                .advanceWatermarkTo(new Instant(16000))
                .advanceProcessingTime(Duration.standardSeconds(2)) // move Processing time to sec 14
                .addElements(
                        TimestampedValue.of("Outside-allowed-lateness", new Instant(9000))

                )
                .advanceWatermarkToInfinity();

        p.apply(streamEvents)
                .apply("KVs", ParDo.of(new DoFn<String, KV<String, String>>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        c.output(KV.of("key", c.element()));
                    }
                }))
                .apply(Window.<KV<String, String>>into(
                        FixedWindows.of(Duration.standardSeconds(windowLength)))
                        .withAllowedLateness(Duration.standardSeconds(allowLateSize))
                        .triggering(
                                // If we dont trigger Repeatedly, after the first trigger, the rest of the elements would be discarded
                                // https://s.apache.org/finishing-triggers-drop-data
                                Repeatedly.forever(
                                        AfterPane.elementCountAtLeast(countElements)
                                                // orFinally stops the repetition of AfterPane
                                                .orFinally(AfterWatermark.pastEndOfWindow())  // Try the pipeline commenting this line!
                                )
                        )
                        .discardingFiredPanes()
                )
                .apply(GroupByKey.create())
                .apply("Log", ParDo.of(new DoFn<KV<String, Iterable<String>>, String>() {
                    @ProcessElement
                    public void processElement(ProcessContext c) {
                        LOG.info("\n TRIGGER " + c.element().getValue().toString());
                        c.output(c.pane().toString());
                    }
                }))
        ;


        p.run();
    }
}
        /*
        EXPLANATION

        The idea of the trigger is to fire every two elements, but there are some constrains. Let's check the output

        1 - The first trigger contains "first" and "second", which is expected since they are the first 2 elements.
        2 - The second trigger contains the third element but all "same-time-x" elements, so it is not just 2 elements.
        This happens because of "same-time-x" arrive at the same time, so there's not second element but second elements
        3 - The third trigger only contains "last-in-window", this is because we have `.orFinally(AfterWatermark.pastEndOfWindow())`
        so it triggers when the window closes, no matter the elements there are.
        4 - The fourth trigger contains "Late-(1,2,3)", since they arrive together
        5 - The next triggers contain the elements as they come, since we have `.orFinally(AfterWatermark.pastEndOfWindow())`
        6 - Element "Outside-allowed-lateness" is discarded as it's outside the allowed lateness

        If we comment ".orFinally(AfterWatermark.pastEndOfWindow())" triggers (3) and (4) are put together, since (3) does not
        contain 2 elements. The next triggers follow the 2 elements rule, so "late-4" and "late-5" are put together, but
        "late-6" is by itself since there are no more elements and the allowed late time passed.

        In a real life scenario, it's not that common two elements arrive at the exact same time.
         */

File: Python/testing_windows/element_count_trigger.py
#  Copyright 2022 Google LLC
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

# standard libraries
import logging

# third party libraries
import apache_beam.transforms.trigger as trigger
from apache_beam import GroupByKey, Map, WindowInto
from apache_beam.options.pipeline_options import PipelineOptions
from apache_beam.testing.test_pipeline import TestPipeline
from apache_beam.testing.test_stream import TestStream
from apache_beam.transforms.window import FixedWindows, TimestampedValue


def get_input_stream():
    stream = (
        TestStream()
        .add_elements(
            [
                # timestamp in seconds since epoch
                TimestampedValue("First", timestamp=0)
            ]
        )
        .advance_watermark_to(2)
        .advance_processing_time(2)
        .add_elements(
            [
                # late from pipeline perspective but still within window
                TimestampedValue("Second", timestamp=1)
            ]
        )
        .advance_watermark_to(4)
        .advance_processing_time(2)
        .add_elements([TimestampedValue("Third", timestamp=3)])
        .advance_watermark_to(6)
        .advance_processing_time(2)
        .add_elements(
            [
                # Different event times but elements "arrived" same time in pipeline
                TimestampedValue("Same-Time-1", timestamp=4),
                TimestampedValue("Same-Time-2", timestamp=5),
                TimestampedValue("Same-Time-3", timestamp=6),
            ]
        )
        .advance_watermark_to(9)
        .advance_processing_time(3)
        .add_elements([TimestampedValue("Last-In-Window", timestamp=8)])
        .advance_watermark_to(11)
        .advance_processing_time(3)
        .add_elements(
            [
                TimestampedValue("Late-1", timestamp=0),
                TimestampedValue("Late-2", timestamp=5),
                TimestampedValue("Late-3", timestamp=6),
            ]
        )
        .advance_watermark_to(12)
        .advance_processing_time(1)
        .add_elements([TimestampedValue("Late-4", timestamp=7)])
        .advance_watermark_to(13)
        .advance_processing_time(1)
        .add_elements([TimestampedValue("Late-5", timestamp=6)])
        .advance_watermark_to(14)
        .advance_processing_time(1)
        .add_elements([TimestampedValue("Late-6", timestamp=0)])
        .advance_watermark_to(15)
        .advance_processing_time(1)
        .add_elements([TimestampedValue("Late-7", timestamp=1)])
        .advance_watermark_to(16)
        .advance_processing_time(1)
        .add_elements(
            [
                TimestampedValue("Outside-Window-Lateness", timestamp=4),
                TimestampedValue("Window-2-In-Time-1", timestamp=14),
            ]
        )
        .advance_watermark_to_infinity()
    )
    return stream


def run():
    options = PipelineOptions(streaming=True)

    p = TestPipeline(options=options)

    window_size_seconds = 10
    window_allowed_lateness_seconds = 5
    late_pane_size = 2

    stream = get_input_stream()

    (
        p
        | stream
        | "Assign elements within pane same key" >> Map(lambda e: ("Key", e))
        | "Window into FixedWindow with triggering"
        >> WindowInto(
            FixedWindows(window_size_seconds),
            allowed_lateness=window_allowed_lateness_seconds,
            accumulation_mode=trigger.AccumulationMode.DISCARDING,
            # Try setting different types of triggers!
            # AfterWatermark only fires pane once when the watermark crosses the
            # Window end time
            # trigger=trigger.AfterWatermark()
            # We  repeatedly fire panes forever with count as least late_pane_size
            # We also want to set a termination to stop above infinite fires so
            # use AfterWatermark a final trigger.
            # trigger=trigger.OrFinally(
            #     trigger.Repeatedly(trigger.AfterCount(late_pane_size)),
            #     trigger.AfterWatermark()
            # )
            # We need to configure how late data should be handled to prevent data
            #  loss since AfterWatermark by default only fires once
            # https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/trigger.py#L596
            trigger=trigger.OrFinally(
                trigger.Repeatedly(trigger.AfterCount(late_pane_size)),
                #   trigger.Repeatedly automatically called for `late` param
                trigger.AfterWatermark(late=trigger.AfterCount(1)),
            ),
        )
        | "Combine elements within pane by key" >> GroupByKey()
        | "Print elements" >> Map(lambda e: print("TRIGGER ", e[1]))
    )

    p.run()


if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    run()

"""
EXPLANATION

`trigger.AfterCount` triggers whenever the current pane has at least `late_pane_size` elements,
 but there are some constrains.

1 - The first trigger contains "First" and "Second", which is expected since they are the first 2 elements.
2 - The second trigger contains the "Third" element but also all of "Same-Time-X" elements, so it is not just 2 elements.
  This happens because of "Same-Time-X" arrive at the same time, so there is no second element but second elements
3 - The third trigger only contains "Last-In-Window", this is because we have `trigger.OrFinally(... , trigger.AfterWatermark(...))`
  so it triggers when the window closes, no matter the number of elements there are in pane.
4 - The fourth trigger contains "Late-(1,2,3)", since they arrive together
5 - The next triggers contain the elements as they come, since we have `trigger.OrFinally(... , trigger.AfterWatermark(...))`
6 - Element "Outside-Window-Lateness" is discarded as it's outside the allowed lateness

In a real life scenario, it's not that common two elements arrive at the exact same time, but networking conditions can affect arrival.
Note: `AfterCount` can cause hanged pipeline if the condition is not met. To resolve this see note in `after_each_trigger.py`
"""
